### PR TITLE
feat: crawler now accepts DNS seeder argument

### PR DESCRIPTION
**Pull request by asmie**
There is now possibility to add DNS seeders as arguments for crawler:
`-d, --dns-seed <DNS_SEED_ADDRS>`
They can be used to obtain IP addresses for initial nodes. DNS seeder can be used instead of local node for crawling as this feature does basically the same - plain node is also checking DNS seeders upon start.

### CODE REVIEW

File: src/tools/crawler/main.rs [hunk 1]
```diff
@@ -75,10 +80,34 @@ fn start_logger(default_level: LevelFilter) {
...
+        seed_addrs.extend_from_slice(&seeds);
+    }
+
+    if let Some(dns_seed) = args.dns_seed {
+        for seed_str in dns_seed {
+            let response = lookup_host(&seed_str);
+
+            if let Ok(response) = response {
+                for address in response.iter() {
+                    seed_addrs.push(SocketAddr::new(*address, 8233)); // DNS addrs use this port
```
Comment (kylegranger) [inline src/tools/crawler/main.rs:98]: maybe make this number a const

**Review by kylegranger — APPROVED** [review]
👍  this should prove to be useful

File: src/tools/crawler/main.rs (hunk 1 above)
Comment (asmie) [reply src/tools/crawler/main.rs]: Right, it's been already defined but wasn't public. 
Changed. 👍

File: src/tools/crawler/main.rs [hunk 2]
```diff
@@ -39,8 +40,12 @@ const LOG_PATH: &str = "crawler-log.txt";
 #[clap(author, version, about, long_about = None)]
 struct Args {
```
Comment (zeapoz) [inline src/tools/crawler/main.rs:41]: How about using an `ArgGroup` to make sure that at least one of the source addrs arguments is given?
https://docs.rs/clap/latest/clap/struct.ArgGroup.html#method.required

This should do the job:
```suggestion
#[clap(author, version, about, long_about = None)]
#[clap(group(
        ArgGroup::new("source_addrs")
            .required(true)
            .args(&["seed-addrs", "dns_seed"])
        ))],
struct Args {
```

**Review by zeapoz — APPROVED** [review]
Looking very good! I can update the crawler workflow once this lands :+1:

File: src/tools/crawler/main.rs (hunk 2 above)
Comment (asmie) [reply src/tools/crawler/main.rs]: Nice one!
Exact form would be like that:

```
#[clap(group(
ArgGroup::new("source_addrs")
.required(true)
.multiple(true)
.args(&["seed_addrs", "dns_seed"])
))]
```

I've added `.multiple(true)` which allows to set both dns seeder and give just peer 😄

File: src/tools/crawler/README.md [hunk 3]
```diff
@@ -12,21 +12,24 @@ $ cargo run --release --features crawler --bin crawler -- --help
...
+            
     -h, --help
             Print help information
 
     -r, --rpc-addr <RPC_ADDR>
             If present, start an RPC server at the specified address
 
     -s, --seed-addrs <SEED_ADDRS>...
-            The initial addresses to connect to
+            The initial addresses to connect to (can be used multiple times)
```
Comment (Rqnsom) [inline src/tools/crawler/README.md:26]: To make the cmd UI simpler (with fewer options) - an alternative solution could have been to allow this argument to receive both (but then we would have used the` Vec<String>` and parsed/checked the IP locally):
 - human-friendly domain names (which we would resolve internally with a DNS service)
 - raw alphanumeric IP address

File: src/tools/crawler/main.rs [hunk 4]
```diff
@@ -37,10 +38,20 @@ const LOG_PATH: &str = "crawler-log.txt";
 
 #[derive(Parser)]
 #[clap(author, version, about, long_about = None)]
+#[clap(group(
+ArgGroup::new("source_addrs")
+.required(true)
+.multiple(true)
+.args(&["seed_addrs", "dns_seed"])
```
Comment (Rqnsom) [inline src/tools/crawler/main.rs:45]: I'm gonna bookmark this part 😃

File: src/tools/crawler/metrics.rs [hunk 5]
```diff
@@ -10,7 +10,7 @@ use crate::{
 };
 
 const MIN_BLOCK_HEIGHT: i32 = 2_000_000;
-const ZCASH_P2P_PORT: u16 = 8233;
+pub const ZCASH_P2P_PORT: u16 = 8233;
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs:13]: the default part implies (for more context) it's not necessarily always the 8233
```suggestion
pub const ZCASH_P2P_DEFAULT_PORT: u16 = 8233;
```

File: src/tools/crawler/main.rs [hunk 6]
```diff
@@ -75,10 +86,34 @@ fn start_logger(default_level: LevelFilter) {
...
+        for seed_str in dns_seed {
+            let response = lookup_host(&seed_str);
+
+            if let Ok(response) = response {
+                for address in response.iter() {
+                    seed_addrs.push(SocketAddr::new(*address, ZCASH_P2P_PORT)); // DNS addrs use this port
+                    info!(parent: crawler.node().span(), "DNS seed {} address added: {}", seed_str, address);
+                }
+            } else {
+                error!(parent: crawler.node().span(), "DNS seed {} lookup failed: {}", seed_str, response.err().unwrap());
```
Comment (Rqnsom) [inline src/tools/crawler/main.rs:108]: Maybe we could kill the process here, so we are aware if this ever happens - otherwise, the mistake might go unnoticed by the user (who might have accidentally entered a wrong domain name)

and the user can be our CI e.g.

File: src/tools/crawler/main.rs [hunk 7]
```diff
@@ -75,10 +86,34 @@ fn start_logger(default_level: LevelFilter) {
...
+            }
+        }
+    }
+
+    if seed_addrs.is_empty() {
+        panic!("Every seed address failed to resolve. Exiting.");
```
Comment (Rqnsom) [inline src/tools/crawler/main.rs:114]: I guess this should never happen
```suggestion
        panic!("seed address not found");
```

**Review by Rqnsom — APPROVED** [review]
I left some ideas, but it looks good - I like it!

File: src/tools/crawler/main.rs (hunk 7 above)
Comment (asmie) [reply src/tools/crawler/main.rs]: Done.

File: src/tools/crawler/metrics.rs (hunk 5 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Good idea! Changed.

File: src/tools/crawler/main.rs (hunk 6 above)
Comment (asmie) [reply src/tools/crawler/main.rs:108]: I'm not convinced to killing process if **any ** seed fails. 
I think that there is quite often use case, when eg. in our CI we'll add 10 DNS seeds (every possible seed). Probably, some of them return the same seeds (as each DNS seed can return many addresses) and it would be ok, when 1 out of 10 would be unavailable - I believe we still would like to crawl the network using remaning 9 seeds.
Constantly changing CI configuration because of one or two unavailable seeds seems to generate more workload to maintain the list than just putting all possible seeds and happily ignore if any of them fails.
Besides, user is still notified by `error` message.

File: src/tools/crawler/README.md (hunk 3 above)
Comment (asmie) [reply src/tools/crawler/README.md:26]: True, that could be a fine feature to recognize and parse different types of addresses. 
There are, however many use cases to handle:
1. user inputs only IP address (need to add default port);
2. user inputs IP address with port (that's current `seed-addrs` case) but need to parse it (now it's done by clap);
3. user inputs hostname - need to perform DNS lookup and append port to each returned IP;
4. user inputs hostname with port - need to perform DNS lookup and use each returned IP with given port.

Basically, DNS seeder return many addresses during lookup and each address points de facto to different host. Plain host given by name should return only one address but it's possible that it returns more than one address (which could point to the same host) - that wouldn't be desired behavior as we'd like get all A records only for DNS seeder not the plain hosts.

That's something worth considering but I'd do it in separate PR as it may involve a bit more coding.

Comment (asmie) [reply src/tools/crawler/README.md:26]: Added to improvement bucket 😄
